topi/include/topi/transform.h

```diff
@@ -358,5 +358,78 @@ inline Array<Tensor> split_sections(const Tensor& x,
   return split(x, split_indices, axis, name, tag);
 }
 
+/*!
+* \brief Take elements from an array along an axis.
+*
+* \param a The source array.
+* \param indices The indices of the values to extract.
+* \param axis The axis over which to select values.
+* \param name The name of the operation.
+* \param tag The tag to mark the operation.
+*
+* \return A Tensor whose op member is the take operation
+*/
+inline Tensor take(const Tensor& a,
+                    const Tensor& indices,
+                    int axis,
+                    std::string name = "tensor",
+                    std::string tag = kInjective) {
+  if (axis < 0) {
+    axis += static_cast<int>(a->shape.size());
+  }
+  CHECK_LT(axis, a->shape.size()) << "axis out of bounds";
+
+  int indices_len = static_cast<int>(indices->shape.size());
+  Array<Expr> out_shape;
+  for (size_t i = 0; i < a->shape.size(); ++i) {
+    if (axis == static_cast<int>(i)) {
+      for (size_t j = 0; j < indices->shape.size(); ++j) {
+        out_shape.push_back(indices->shape[j]);
+      }
+    } else {
+      out_shape.push_back(a->shape[i]);
+    }
+  }
+  return compute(
+        out_shape, [&](const Array<Var>& out_index) {
+          Array<Expr> indices_position;
+          for (size_t j = axis; j < static_cast<size_t>(axis+indices_len); ++j) {
+            indices_position.push_back(out_index[j]);
+          }
+          auto index_along_axis = indices(indices_position);
+          Array<Expr> real_indices;
+          for (size_t j = 0; j < static_cast<size_t>(axis); ++j) {
+            real_indices.push_back(out_index[j]);
+          }
+          real_indices.push_back(index_along_axis);
+          for (size_t j = axis + indices_len; j < out_index.size(); ++j) {
+            real_indices.push_back(out_index[j]);
+          }
+          return a(real_indices);
+        }, name, tag);
+}
+
+  /*!
+  * \brief Take elements from an flattened input array when axis is None.
+  *
+  * \param a The source array.
+  * \param indices The indices of the values to extract.
+  * \param name The name of the operation.
+  * \param tag The tag to mark the operation.
+  *
+  * \return A Tensor whose op member is the take operation
+  */
+inline Tensor take_flatten(const Tensor& a,
+                           const Tensor& indices,
+                           std::string name = "tensor",
+                           std::string tag = kInjective) {
+  int a_size = 1;
+  for (size_t posi = 0; posi < a->shape.size(); ++posi) {
+    a_size = a_size * static_cast<int>(GetConstInt(a->shape[posi]));
+  }
+  Array<Expr> a_shape_exp;
+  a_shape_exp.push_back(a_size);
+  return take(reshape(a, a_shape_exp), indices, 0);
```

Comment:
Let us also expland the compute here, since it is not hard to implement a flattened reference to a. Reuse the index implementation of flatten